File: String/49_GroupAnagrams.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        ht = defaultdict(list)
        for str in strs:
            key = "".join(sorted(str))
            ht[key].append(str)
        return list(ht.values())


class Solution2:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        ht = defaultdict(list)
        for str in strs:
            cnt = [0] * 26
            for c in str:
                cnt[ord(c) - ord("a")] += 1
            ht[tuple(cnt)].append(str)
        return list(ht.values())

File: Java Solutions/49_GroupAnagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> map = new HashMap<>();
        for (String word : strs) {
            char[] chars = word.toCharArray();
            Arrays.sort(chars);
            String sortedWord = new String(chars);
            if (!map.containsKey(sortedWord)) {
                map.put(sortedWord, new ArrayList<>());
            }
            map.get(sortedWord).add(word);
        }
        return new ArrayList<>(map.values());
    }
}


class Solution2 {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> map = new HashMap<>();
        for (String word : strs) {
            int[] counter = new int[26];
            for (char ch : word.toCharArray()) {
                counter[ch - 'a']++;
            }
            StringBuilder sb = new StringBuilder();
            for (int i = 0; i < 26; i++) {
                if (counter[i] != 0) {
                    sb.append((char) ('a' + i));
                    sb.append(counter[i]);
                }
            }
            String key = sb.toString();
            List<String> list = map.getOrDefault(key, new ArrayList<String>());
            list.add(word);
            map.put(key, list);
        }
        return new ArrayList<>(map.values());
    }
}